src/backend/distributed/commands/multi_copy.c

```diff
@@ -3571,6 +3613,52 @@ CopyGetPlacementConnection(HTAB *connectionStateHash, ShardPlacement *placement,
 	}
 
 	connection = GetPlacementConnection(connectionFlags, placement, nodeUser);
+	if (connection == NULL)
+	{
+		/*
+		 * For this COPY command, we have not established any connections and adaptive
+		 * connection management throttled the new connection request. This could only
+		 * happen if this COPY command is the second (or later) COPY command in a
+		 * transaction block as the first COPY command always gets a connection per
+		 * node thanks to the connection reservation.
+		 *
+		 * As we know that there has been at least one COPY command happened earlier,
+		 * we need to find the connection to that node, and use it.
+		 */
+		if (list_length(copyConnectionStateList) == 0)
+		{
+			connection =
+				ConnectionAvailableToNode(nodeName, nodePort, CurrentUserName(),
+										  CurrentDatabaseName());
+
+			/*
+			 * We do not expect this to happen, but still instead of an assert,
+			 * we prefer explicit error message.
+			 */
+			if (connection == NULL)
+			{
+				ereport(ERROR, (errmsg("unexpected connection state. This is "
+									   "probably a Citus bug so please report it"),
```

Comment:
maybe "unexpected connection state" -> "could not find an available connection"

Could be nice to move the "This is probably a Citus bug ..." to the errdetail, mostly because the inconsistent punctuation ticks me off :).